distributed/cluster_dump.py

```diff
@@ -57,3 +58,112 @@ def writer(state: dict, f: IO):
         # Write from a thread so we don't block the event loop quite as badly
         # (the writer will still hold the GIL a lot though).
         await to_thread(writer, state, f)
+
+
+def load_cluster_dump(url: str):
+    if url.endswith(".msgpack.gz"):
+        mode = "rb"
+        reader = msgpack.unpack
+    elif url.endswith(".yaml"):
+        import yaml
+
+        mode = "r"
+        reader = yaml.safe_load
+    else:
+        raise ValueError(f"url ({url}) must have a .msgpack.gz or .yaml suffix")
+
+    with fsspec.open(url, mode, compression="infer") as f:
+        return reader(f)
+
+
+class DumpInspector:
+    """
+    Utility class for inspecting the state of a cluster dump
+
+    .. code-block:: python
+
+        inspector = DumpInspect("dump.msgpack.gz")
+        memory_tasks = inspector.tasks_in_state("memory")
+        released_tasks = inspector.tasks_in_state("released)
+    """
+
+    def __init__(self, url_or_state: str | dict):
+        if isinstance(url_or_state, str):
+            self.dump = load_cluster_dump(url_or_state)
+        elif isinstance(url_or_state, dict):
+            self.dump = url_or_state
+        else:
+            raise TypeError("'url_or_state' must be a str or dict")
+
+    def tasks_in_state(self, state: str = "", workers: bool = False) -> dict:
+        """
+        Returns
+        -------
+        tasks : dict
+            A dictionary of scheduler tasks with state `state`.
+            worker tasks are included if `workers=True`
+        """
+        stasks = self.dump["scheduler"]["tasks"]
+
+        if state:
+            tasks = {k: v for k, v in stasks.items() if v["state"] == state}
+        else:
+            tasks = stasks.copy()
+
+        if not workers:
+            return tasks
+
+        for worker_dump in self.dump["workers"].values():
+            if self._valid_worker_dump(worker_dump):
+                if state:
+                    tasks.update(
+                        (k, v)
+                        for k, v in worker_dump["tasks"].items()
+                        if v["state"] == state
+                    )
+                else:
+                    tasks.update(worker_dump["tasks"])
```

Comment:
I'm wondering how the output format of this looks like. I suggest to add an example to the docs or add a test with an example.

I'm wondering if we should mix scheudler and worker keys here. If the format makes it easy enough to distinguish it. There is also the option to introduce two different fucntions for scheduler and worker

Reply:
> I'm wondering how the output format of this looks like. I suggest to add an example to the docs or add a test with an example.

It returns a dictionary of tasks, keyed on task-id. I think an example in the docs might not work too well as the task dictionaries are rather large -- I think it would make more sense to add a test in this case.
 
> I'm wondering if we should mix scheudler and worker keys here. If the format makes it easy enough to distinguish it. There is also the option to introduce two different fucntions for scheduler and worker

On reflection, I'm also concerned as to whether mixing scheduler and worker tasks is a good idea. In the worst case, could there be records of the same task on the scheduler and multiple workers? In that case, a simple dictionary keyed on task-id might inadvertently destroy useful information.

If so, I can see a couple of alternatives:

1. Separate functions for the scheduler and workers (and to maintain a coherent interface, we could do the same with stories)
2.  Return a `defaultdict(list)` of tasks keyed on task id.

Reply:
Based on our discussion yesterday, I've split the task and story retrieval functionality into scheduler and worker functions. Additionally, these functions simply return lists of dicts, which are suitable for import into pandas